apps/emqx_dashboard/src/emqx_dashboard.erl

```diff
@@ -10,43 +10,50 @@
     stop_listeners/1,
     stop_listeners/0,
     list_listeners/0,
-    wait_for_listeners/0
+    listeners_status/0,
+    regenerate_dispatch_after_config_update/0
 ]).
 
 %% Authorization
 -export([authorize/2]).
 
--export([save_dispatch_eterm/1, save_dispatch_eterm/2]).
-
 -include_lib("emqx/include/logger.hrl").
 -include_lib("emqx/include/http_api.hrl").
 -include_lib("emqx/include/emqx_release.hrl").
--dialyzer({[no_opaque, no_match, no_return], [init_cache_dispatch/2, start_listeners/1]}).
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
 
 -define(EMQX_MIDDLE, emqx_dashboard_middleware).
--define(DISPATCH_FILE, "dispatch.eterm").
+
+-type listener_name() :: atom().
+-type listener_configs() :: #{listener_name() => emqx_config:config()}.
+
+-export_type([listener_name/0, listener_configs/0]).
 
 %%--------------------------------------------------------------------
 %% Start/Stop Listeners
 %%--------------------------------------------------------------------
 
+-spec start_listeners() -> ok | {error, [listener_name()]}.
 start_listeners() ->
     start_listeners(listeners()).
 
+-spec stop_listeners() -> ok.
 stop_listeners() ->
     stop_listeners(listeners()).
 
+-spec start_listeners(listener_configs()) ->
+    ok | {error, [listener_name()]}.
 start_listeners(Listeners) ->
-    {ok, _} = application:ensure_all_started(minirest),
-    SwaggerSupport = emqx:get_config([dashboard, swagger_support], true),
-    InitDispatch = dispatch(),
+    %% NOTE
+    %% Before starting the listeners, we do not generate the full dispatch upfront,
+    %% because the listeners may fail to start and the generation may appear useless.
+    InitDispatch = init_dispatch(),
```

Comment:
Just to discuss. As an occasional EMQX user / operator / integration tester, I'd prefer EMQX Dashboard listener available = EMQX API usable. As far as I understand, this is not exactly the case right now because of this short phase where the listener starts with not-yet-initialized dispatch. Now that this "regeneration" is much more effective, this sort of optimization also feels a bit unnecessary, especially since we don't really expect listeners to fail to start.